### PR TITLE
Add Non-Faulting behaviour for Sve.LoadVector*NonFaulting*()

### LINKED ISSUE #105163: JIT: SVE - Add FFR lifetime tracking as well as test coverage for NonFaulting APIs
Since we have already merged the NonFaulting SVE APIs, we need to add additional testing on their first-faulting behavior since they share the similar behavior as the FirstFaulting APIs. See https://github.com/dotnet/runtime/pull/104502

### PR DESCRIPTION
**Pull request by SwapnilGaikwad**
Fixes: #105163

It also fixes the issue where FFR was not preserved and side-effects from the intrinsic not specified.

@a74nh @kunalspathak @dotnet/arm64-contrib @arch-arm64-sve @TIHan @amanasifkhalid

### CODE REVIEW

**Comment by SwapnilGaikwad** [comment]
<details>
<summary>All stress tests are passing.</summary>

```
===================Running default===================
------------------- {} -------------------
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_float() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_double() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_sbyte() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_short() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_int() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_long() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_byte() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_ushort() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_uint() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorNonFaulting_ulong() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorInt16NonFaultingSignExtendToInt32_int() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorInt16NonFaultingSignExtendToInt64_long() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorInt16NonFaultingSignExtendToUInt32_uint() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorInt16NonFaultingSignExtendToUInt64_ulong() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorInt32NonFaultingSignExtendToInt64_long() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorInt32NonFaultingSignExtendToUInt64_ulong() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorSByteNonFaultingSignExtendToInt16_short() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorSByteNonFaultingSignExtendToInt32_int() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorSByteNonFaultingSignExtendToInt64_long() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorSByteNonFaultingSignExtendToUInt16_ushort() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorSByteNonFaultingSignExtendToUInt32_uint() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorSByteNonFaultingSignExtendToUInt64_ulong() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorByteNonFaultingZeroExtendToInt16() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorByteNonFaultingZeroExtendToInt32() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorByteNonFaultingZeroExtendToInt64() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorByteNonFaultingZeroExtendToUInt16() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorByteNonFaultingZeroExtendToUInt32() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorByteNonFaultingZeroExtendToUInt64() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorUInt16NonFaultingZeroExtendToInt32() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorUInt16NonFaultingZeroExtendToInt64() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorUInt16NonFaultingZeroExtendToUInt32() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorUInt16NonFaultingZeroExtendToUInt64() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorUInt32NonFaultingZeroExtendToInt64() : 18
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_LoadVectorUInt32NonFaultingZeroExtendToUInt64() : 18
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```
</details>

**Review by a74nh — APPROVED** [review]
This LGTM.

Looks like we are missing FFR tracking for gatherload first faulting methods too. We should do that in a follow on PR.

**Comment by SwapnilGaikwad** [comment]
> This LGTM.
> 
> Looks like we are missing FFR tracking for gatherload first faulting methods too. We should do that in a follow on PR.

As discussed offline, the `LoadVectorNonFaulting` are in SVE2 so we can skip them for now🙂.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/SveLoadNonFaultingUnOpTest.template [hunk 1]
```diff
@@ -81,6 +82,9 @@ namespace JIT.HardwareIntrinsics.Arm
             {
                 int sizeOfinArray1 = inArray1.Length * Unsafe.SizeOf<{Op1BaseType}>();
                 int sizeOfoutArray = outArray.Length * Unsafe.SizeOf<{RetBaseType}>();
+                // Choose size of bounded memory such that half of the vector would fall in accessible memory while the rest would in non-accesible one.
```
Comment (amanasifkhalid) [inline src/tests/JIT/HardwareIntrinsics/Arm/Shared/SveLoadNonFaultingUnOpTest.template:85]:
```suggestion
                // Choose size of bounded memory such that half of the vector would fall in accessible memory while the rest would in non-accessible one.
```

**Review by amanasifkhalid — APPROVED** [review]
LGTM too, thanks!

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/SveLoadNonFaultingUnOpTest.template (hunk 1 above)
Comment (SwapnilGaikwad) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/SveLoadNonFaultingUnOpTest.template]: Done.

**Comment by amanasifkhalid** [comment]
/ba-g runtime-dev-innerloop failure is from CI machine rebooting

**Comment by amanasifkhalid** [comment]
> Looks like we are missing FFR tracking for gatherload first faulting methods too. We should do that in a follow on PR.

@SwapnilGaikwad are you planning on doing this item? I think this is the source of the failure in #106621.

**Comment by SwapnilGaikwad** [comment]
> > Looks like we are missing FFR tracking for gatherload first faulting methods too. We should do that in a follow on PR.
> 
> @SwapnilGaikwad are you planning on doing this item? I think this is the source of the failure in #106621.

Hi @amanasifkhalid, there were `GatherVectorFirstFaulting` and `LoadVectorFirstFautling` in SVE1. We have the FFR tracking for them. Alan may have referred to APIs from SVE2 group so skipped them.
However, happy to take a look at the #106621 👍

**Comment by SwapnilGaikwad** [comment]
Continuing conversation [here](https://github.com/dotnet/runtime/issues/106621#issuecomment-2304649196).

**Pull request merged**